From 8c8c110a934a9abd7e69762f89f4029bea7f9c4a Mon Sep 17 00:00:00 2001
From: Harald Sitter <sitter@kde.org>
Date: Mon, 24 Oct 2022 13:18:20 +0200
Subject: [PATCH] disable automatic portal launching early on

this opts out of color picking support early on in the login process
when our environment is still incomplete. not doing this would cause the
portal system to launch "too early" and have an incomplete environment
set. instead disable the portal support early on and only enable it once
we have a working environment constructed.

also move cursor setup and ksplash into this newly available "critical
section". now that we have an opt out we can run them as early as
necessary without risking them starting the portal infrastructure.

BUG: 458865


(cherry picked from commit 670cf731a75037c646864ed0bd03a35a35130c5e)
---
 startkde/startplasma-x11.cpp | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/startkde/startplasma-x11.cpp b/startkde/startplasma-x11.cpp
index 08a4c0d1c0..0e20300de5 100644
--- a/startkde/startplasma-x11.cpp
+++ b/startkde/startplasma-x11.cpp
@@ -21,6 +21,7 @@ int main(int argc, char **argv)
     // When the X server dies we get a HUP signal from xinit. We must ignore it
     // because we still need to do some cleanup.
     signal(SIGHUP, sighupHandler);
+    qputenv("QT_NO_XDG_DESKTOP_PORTAL", QByteArrayLiteral("1"));
 
     QCoreApplication app(argc, argv);
 
@@ -57,6 +58,11 @@ int main(int argc, char **argv)
         }
     }
 
+    // NOTE: Be very mindful of what you start this early in the process. The environment is not yet complete.
+    setupCursor(false);
+    std::unique_ptr<QProcess, KillBeforeDeleter> ksplash(setupKSplash());
+    Q_UNUSED(ksplash)
+
     runEnvironmentScripts();
 
     out << "startkde: Starting up...\n";
@@ -64,6 +70,7 @@ int main(int argc, char **argv)
     setupPlasmaEnvironment();
     setupX11();
 
+    qunsetenv("QT_NO_XDG_DESKTOP_PORTAL");
     auto oldSystemdEnvironment = getSystemdEnvironment();
     if (!syncDBusEnvironment()) {
         // Startup error
@@ -76,12 +83,6 @@ int main(int argc, char **argv)
     // variables (e.g. LANG and LC_*)
     importSystemdEnvrionment();
 
-    // NOTE: Do not start QGuiApplications before setting up the environment. We'd be at risk of dbus invoking other
-    // processes with an incomplete environment.
-    setupCursor(false);
-    std::unique_ptr<QProcess, KillBeforeDeleter> ksplash(setupKSplash());
-    Q_UNUSED(ksplash)
-
     if (!startPlasmaSession(false))
         return 1;
 
-- 
GitLab

